Panic due to null math

Repro is with Zed commit 19b2eb5.

```
$ zq -version
Version: v1.14.0-29-g19b2eb5d

$ zq 'yield null +1'
panic: null
goroutine 1 [running]:
runtime/debug.Stack()
	/usr/local/opt/go@1.21/libexec/src/runtime/debug/stack.go:24 +0x5e
github.com/brimdata/zed/runtime/sam/op.(*Catcher).Pull.func1()
	/Users/phil/work/zed/runtime/sam/op/catcher.go:25 +0x3d
panic({0x2360980?, 0xc000557430?})
	/usr/local/opt/go@1.21/libexec/src/runtime/panic.go:914 +0x21f
github.com/brimdata/zed/runtime/sam/expr/coerce.ToNumeric[...]({{0x28c6610?, 0x36dcd00?}, 0x0?, 0xc00052cf40?})
	/Users/phil/work/zed/runtime/sam/expr/coerce/coerce.go:53 +0x125
github.com/brimdata/zed/runtime/sam/expr.toInt(...)
	/Users/phil/work/zed/runtime/sam/expr/eval.go:355
github.com/brimdata/zed/runtime/sam/expr.(*Add).Eval(0xc000571230, {0x4ab35c48?, 0xc00052cf40?}, {{0x28c6610?, 0x36dcd00?}, 0x0?, 0x1011285?})
	/Users/phil/work/zed/runtime/sam/expr/eval.go:412 +0x385
github.com/brimdata/zed/runtime/sam/op/yield.(*Op).Pull(0xc000571260, 0x5?)
	/Users/phil/work/zed/runtime/sam/op/yield/yield.go:31 +0x432
github.com/brimdata/zed/runtime/sam/op.(*Single).Pull(0xc000510e70, 0x20)
	/Users/phil/work/zed/runtime/sam/op/mux.go:134 +0x2b
github.com/brimdata/zed/runtime/sam/op.(*Catcher).Pull(0x3cbfa68?, 0x10?)
	/Users/phil/work/zed/runtime/sam/op/catcher.go:28 +0x62
github.com/brimdata/zed/runtime/exec.(*Query).Pull(0x4ab35b10?, 0xd8?)
	/Users/phil/work/zed/runtime/exec/query.go:59 +0x3e
github.com/brimdata/zed/zbuf.CopyPuller({0x4ab35b10, 0xc000510cf0}, {0x28c0420, 0xc000571290})
	/Users/phil/work/zed/zbuf/batch.go:162 +0x4d
github.com/brimdata/zed/cli/zq.(*Command).Run(0xc00012ea80, {0xc0000dc010, 0x1, 0x1})
	/Users/phil/work/zed/cli/zq/command.go:162 +0x79e
github.com/brimdata/zed/pkg/charm.path.run({0xc0000c2350?, 0x1, 0x1}, {0xc0000dc010?, 0x1, 0x0?})
	/Users/phil/work/zed/pkg/charm/path.go:11 +0x79
github.com/brimdata/zed/pkg/charm.(*Spec).ExecRoot(0x100874b?, {0xc0000dc010, 0x1, 0x1})
	/Users/phil/work/zed/pkg/charm/charm.go:63 +0x3f
main.main()
	/Users/phil/work/zed/cmd/zq/main.go:11 +0x5b
```

I confirmed via binary search that that commit 19b2eb5 (associated with the changes in #4719) was when this panic started. At the prior commit:

```
$ zq -version
Version: v1.14.0-28-g05fdd90a

$ zq 'yield null +1'
1
```

--- a/runtime/sam/expr/coerce/coerce.go
+++ b/runtime/sam/expr/coerce/coerce.go
@@ -41,6 +41,9 @@ func Equal(a, b zed.Value) bool {
 }
 
 func ToNumeric[T constraints.Integer | constraints.Float](val zed.Value) T {
+	if val.IsNull() {
+		return 0
+	}
 	switch id := val.Type().ID(); {
 	case zed.IsUnsigned(id):
 		return T(val.Uint())

--- a/runtime/sam/expr/eval.go
+++ b/runtime/sam/expr/eval.go
@@ -357,6 +357,7 @@ func compareNumbers(a, b zed.Value, aid, bid int) int {
 
 func toFloat(val zed.Value) float64 { return coerce.ToNumeric[float64](val) }
 func toInt(val zed.Value) int64     { return coerce.ToNumeric[int64](val) }
+func toUint(val zed.Value) uint64   { return coerce.ToNumeric[uint64](val) }
 
 type Add struct {
 	zctx     *zed.Context
@@ -411,7 +412,7 @@ func (a *Add) Eval(ectx Context, this zed.Value) zed.Value {
 	}
 	switch id := typ.ID(); {
 	case zed.IsUnsigned(id):
-		return zed.NewUint(typ, lhsVal.Uint()+rhsVal.Uint())
+		return zed.NewUint(typ, toUint(lhsVal)+toUint(rhsVal))
 	case zed.IsSigned(id):
 		return zed.NewInt(typ, toInt(lhsVal)+toInt(rhsVal))
 	case zed.IsFloat(id):
@@ -430,7 +431,7 @@ func (s *Subtract) Eval(ectx Context, this zed.Value) zed.Value {
 	}
 	switch id := typ.ID(); {
 	case zed.IsUnsigned(id):
-		return zed.NewUint(typ, lhsVal.Uint()-rhsVal.Uint())
+		return zed.NewUint(typ, toUint(lhsVal)-toUint(rhsVal))
 	case zed.IsSigned(id):
 		if id == zed.IDTime {
 			// Return the difference of two times as a duration.
@@ -450,7 +451,7 @@ func (m *Multiply) Eval(ectx Context, this zed.Value) zed.Value {
 	}
 	switch id := typ.ID(); {
 	case zed.IsUnsigned(id):
-		return zed.NewUint(typ, lhsVal.Uint()*rhsVal.Uint())
+		return zed.NewUint(typ, toUint(lhsVal)*toUint(rhsVal))
 	case zed.IsSigned(id):
 		return zed.NewInt(typ, toInt(lhsVal)*toInt(rhsVal))
 	case zed.IsFloat(id):
@@ -466,11 +467,11 @@ func (d *Divide) Eval(ectx Context, this zed.Value) zed.Value {
 	}
 	switch id := typ.ID(); {
 	case zed.IsUnsigned(id):
-		v := rhsVal.Uint()
+		v := toUint(rhsVal)
 		if v == 0 {
 			return d.zctx.NewError(ectx.Arena(), DivideByZero)
 		}
-		return zed.NewUint(typ, lhsVal.Uint()/v)
+		return zed.NewUint(typ, toUint(lhsVal)/v)
 	case zed.IsSigned(id):
 		v := toInt(rhsVal)
 		if v == 0 {
@@ -494,7 +495,7 @@ func (m *Modulo) Eval(ectx Context, this zed.Value) zed.Value {
 	}
 	switch id := typ.ID(); {
 	case zed.IsUnsigned(id):
-		v := rhsVal.Uint()
+		v := toUint(rhsVal)
 		if v == 0 {
 			return m.zctx.NewError(ectx.Arena(), DivideByZero)
 		}

--- a/runtime/sam/expr/expr_test.go
+++ b/runtime/sam/expr/expr_test.go
@@ -363,6 +363,16 @@ func TestArithmetic(t *testing.T) {
 	testSuccessful(t, "this / 2", val, "5")
 	testSuccessful(t, "this % 3", val, "1")
 
+	// Test arithmetic with null values
+	testSuccessful(t, "null + 1", "", "1")
+	testSuccessful(t, "uint64(1) + null", "", "1(uint64)")
+	testSuccessful(t, "null + 1.0", "", "1.")
+	testSuccessful(t, "1. - null", "", "1.")
+	testSuccessful(t, "uint64(1) * null", "", "0(uint64)")
+	testSuccessful(t, "null / 1.", "", "0.")
+	testSuccessful(t, "1 / uint64(null)", "", `error("divide by zero")`)
+	testSuccessful(t, "null % 1", "", "0")
+
 	// Difference of two times is a duration
 	testSuccessful(t, "a - b", "{a:2022-09-22T00:00:01Z,b:2022-09-22T00:00:00Z}", "1s")
 

